libyul/optimiser/FunctionSpecializer.h

```diff
@@ -0,0 +1,104 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/ASTWalker.h>
+
+#include <libyul/ASTForward.h>
+#include <libyul/Dialect.h>
+
+#include <optional>
+#include <vector>
+#include <map>
+
+namespace solidity::yul
+{
+
+/**
+ * FunctionSpecializer: Optimiser step that specializes the function with its literal arguments.
+ *
+ * If a function, say, `function f(a, b) { sstore (a, b)}`, is called with literal arguments, for
+ * example, `f(x, 5)`, where `x` is an identifier, it could be specialized by creating a new
+ * function `f_1` that takes only one argument, i.e.,
+ *
+ *   function f_1(a_1) {
+ *     let b_1 := 5
+ *     sstore(a_1, b_1)
+ *   }
+ *
+ * Other optimization steps will be able to make more simplifications to the function. The
+ * optimization step is mainly useful for functions that would not be inlined.
+ *
+ * Prerequisites: Disambiguator, FunctionHoister
+ *
+ * LiteralRematerialiser is recommended as a prerequisite, even though it's not required for
+ * correctness.
+ */
+class FunctionSpecializer: public ASTModifier
+{
+public:
+	/// A vector of function-call arguments. An element 'has value' if it's a literal, and the
+	/// corresponding Expression would be the literal.
+	using LiteralArguments = std::vector<std::optional<Expression>>;
+
+	static constexpr char const* name{"FunctionSpecializer"};
+	static void run(OptimiserStepContext& _context, Block& _ast);
+
+	using ASTModifier::operator();
+	virtual void operator()(FunctionCall& _f) override;
+
+private:
+	explicit FunctionSpecializer(NameDispenser& _nameDispenser, Dialect const& _dialect):
+		m_nameDispenser(_nameDispenser),
+		m_dialect(_dialect)
+	{}
+	/// Returns a vector of Expressions, where the index `i` is an expression if the function's
+	/// `i`-th argument can be specialized.
+	LiteralArguments specializableArguments(FunctionCall const& _f);
+	/// Given a function definition `_f` and its arguments `_arguments`, of which, at least one is a
+	/// literal, this function returns a new function with the literal arguments specialized.
+	///
+	/// Note that the returned function definition will have new (and unique) names, for both the
+	/// function and variable declarations to retain the properties enforced by the Disambiguator.
+	///
+	/// For example, if `_f` is the function `function f(a, b, c) -> d { sstore(a, b) }` and
+	/// `_arguments` is the vector of literals `{1, 2, nullopt}`. The returned function could be:
+	///
+	///   function f_1(c_2) -> d_3 {
+	///     let a_4 := 1
+	///     let b_5 := 2
+	///     sstore(a_4, b_5)
+	///   }
+	FunctionDefinition specialize(
+		FunctionDefinition const& _f,
+		YulString _newName,
```

Comment:
Is `_newName` the new name assigned to the specialised function i.e., `f_1` in the example?

Reply:
If so, perhaps add this to the documentation?

Reply:
Added this.